libs/utils/src/sync/gate.rs

```diff
@@ -1,83 +1,101 @@
-use std::{sync::Arc, time::Duration};
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
 
 /// Gates are a concurrency helper, primarily used for implementing safe shutdown.
 ///
 /// Users of a resource call `enter()` to acquire a GateGuard, and the owner of
 /// the resource calls `close()` when they want to ensure that all holders of guards
 /// have released them, and that no future guards will be issued.
 pub struct Gate {
-    /// Each caller of enter() takes one unit from the semaphore. In close(), we
-    /// take all the units to ensure all GateGuards are destroyed.
-    sem: Arc<tokio::sync::Semaphore>,
-
-    /// For observability only: a name that will be used to log warnings if a particular
-    /// gate is holding up shutdown
-    name: String,
+    inner: Arc<GateInner>,
 }
 
 impl std::fmt::Debug for Gate {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "Gate<{}>", self.name)
+        f.debug_struct("Gate")
+            // use this for identification
+            .field("ptr", &Arc::as_ptr(&self.inner))
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+struct GateInner {
+    sem: tokio::sync::Semaphore,
+    closing: std::sync::atomic::AtomicBool,
```

Comment:
Since we log at INFO level immediately before setting this atomic, is it really useful?  When we're logging a gate, aren't we usually logging it _because_ it's slow during close, so we already know it's closing?

Reply:
Oh, this is used by the GateGuard - I see: we're logging who the slow GateGuards were, but only if they eventually complete

Reply:
I am unclear on what are you proposing? Renaming the `closing` or changing how the logging works?

Reply:
Yes, this will not allow us to catch for example `std::mem::forget` situations, but it doesn't seem like we have them now.

Reply:
In this context I was only worried about identifying gates when their guards block forever, but we're covering that on another thread :+1: